### DISCUSSION TITLE
Grain Extensions and their own state

### DISCUSSION THREAD

**Discussion by SGStino** [dotnet/orleans#8795] [Q&A]
I'm trying to write an extension that can be added to a grain that is relatively independant of the grain itself, but it has state. State that the grain in and on it's own, doesn't strictly need to know about.

In theory, I could implement it as a separate grain with a compound key of the parent's primary key and grain type to avoid collisions for this use case, but i feel like extensions would enable some composition of functionality in grains?

**Example use case:**
> In a game: a grain is an entity/gameobject, it can be a player, an interactable object, an npc, anything really..
> And each entity can have some sort of inventory containing items.
> 
> An extension could be written that interacts with that inventory state, but the entity's individual state isn't required to know about the inventory's state. In the persistence layer, it can very well be a different table with same primary key.

The way i interpret the documentation now, is that I should construct the extension component in the OnActivate of the grain, and pass it a reference (or getter/setter) to a state object from the grain, but that would require the grain to have knowledge about the functionality provided by the extension, coupling it a little tighter than i'd like them to be.

Am I trying to abuse something for what it wasn't designed/intended, or is this a legitimate use case that wasn't considered yet?

**Comment by pjmagee** [comment]
Not really an answer, but I'm struggling to understand how to use the Extensions properly myself. All the samples in the dotnet repo only have simple grain examples with simple interfaces. And the examples on the msdocs also are too basic for me to fully understand how to utilise them or if I am going to make something wrong with extensions. Sorry, not really an answer but thought i'd share youre not the only one not really understanding how to create or use a grain extension, as the documentation is extremely light and there's really no samples.

**Reply by SGStino** [reply]
Yes, that's pretty much my sentiment on the extensions, they feel like they can do a lot, on paper, but when actually trying to do something functional, i'm always hitting a wall. Most likely because i don't knowhere the ladder is hidden, but still.
